ODIN_II/SRC/memories.cpp

```diff
@@ -1765,3 +1897,471 @@ signal_list_t* create_decoder(nnode_t* node, short mark, signal_list_t* input_li
+    int offset = 0;
+    /* creating a new node */
+    sp_ram_signals* signals = init_sp_ram_signals();
+
+    /* INPUTS */
+    /* adding the addr signals */
+    for (i = 0; i < SP_ADDR_width; i++) {
+        npin_t* pin = node->input_pins[offset + i];
+        /* detach from the main node, since it will be connected to a new dpram */
+        pin->node->input_pins[pin->pin_node_idx] = NULL;
+
+        add_pin_to_signal_list(signals->addr, pin);
+    }
+    offset += SP_ADDR_width;
+
+    /* adding the clk signals */
+    npin_t* clk_pin = node->input_pins[offset];
+    /* detach from the main node, since it will be connected to a new dpram */
+    clk_pin->node->input_pins[clk_pin->pin_node_idx] = NULL;
+    signals->clk = clk_pin;
+    /* increment offset */
+    offset += 1;
+
+    /* adding the data signals */
+    for (i = 0; i < SP_DATA_width; i++) {
+        /* hook the data1 pin to new node */
+        npin_t* pin = node->input_pins[offset + i];
+        /* in case of padding, pins have not been remapped, need to detach them from the BRAM node */
+        pin->node->input_pins[pin->pin_node_idx] = NULL;
+
+        add_pin_to_signal_list(signals->data, pin);
+    }
+    offset += SP_DATA_width;
+
+    /* adding the we signals */
+    npin_t* we_pin = node->input_pins[offset];
+    /* detach from the main node, since it will be connected to a new dpram */
+    we_pin->node->input_pins[we_pin->pin_node_idx] = NULL;
+    signals->we = we_pin;
+
+    /* OUTPUT */
+    /* adding the output signals */
+    offset = 0;
+    for (i = 0; i < SP_OUT_width; i++) {
+        /* hook the data1 pin to new node */
+        npin_t* pin = node->output_pins[offset + i];
+        /* in case of padding, pins have not been remapped, need to detach them from the BRAM node */
+        pin->node->output_pins[pin->pin_node_idx] = NULL;
+
+        add_pin_to_signal_list(signals->out, pin);
+    }
+
+    /* creating a new spram with size modified input signals */
+    nnode_t* spram = create_single_port_ram(signals, node);
+    /* adding new spram new dp memory list for future iterations and adjusments */
+    sp_memory_list = insert_in_vptr_list(sp_memory_list, spram);
+    /* register the single port ram in arch model to have the related model (SPRAM) in BLIF for simulation */
+    register_memory_model(spram);
+
+    // CLEAN UP
+    free_nnode(node);
+    free_sp_ram_signals(signals);
+}
+
+/**
+ * (function: resolve_dual_port_ram)
+ * 
+ * @brief resolve the dpram block by reordering the input signals 
+ * to be compatible with Odin's partial mapping phase
+ * 
+ * @param node pointing to a dual port ram node
+ * @param traverse_mark_number unique traversal mark for blif elaboration pass
+ * @param netlist pointer to the current netlist file
+ */
+void resolve_dual_port_ram(nnode_t* node, uintptr_t traverse_mark_number, netlist_t* netlist) {
+    oassert(node->traverse_visited == traverse_mark_number);
```

Comment:
Could any user input ever trigger these asserts (without an odinII bug)?  If so, they should print error messages first.